optuna/integration/botorch.py

```diff
@@ -0,0 +1,467 @@
+
+        best_f = (train_obj * (train_con <= 0)).max()
+
+        constraints = []
+        n_contraints = train_con.size(1)
+        for i in range(n_contraints):
+            constraints.append(lambda Z, i=i: Z[..., -n_contraints + i])
+        objective = ConstrainedMCObjective(
+            objective=lambda Z: Z[..., 0],
+            constraints=constraints,
+        )
+    else:
+        train_y = train_obj
+
+        best_f = train_obj.max()
+
+        objective = None  # Using the default identity objective.
+
+    train_x = normalize(train_x, bounds=bounds)
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.size(-1)))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_model(mll)
+
+    acqf = qExpectedImprovement(
+        model=model,
+        best_f=best_f,
+        sampler=SobolQMCNormalSampler(num_samples=256),
+        objective=objective,
+    )
+
+    standard_bounds = torch.zeros_like(bounds)
+    standard_bounds[1] = 1
+
+    candidates, _ = optimize_acqf(
+        acq_function=acqf,
+        bounds=standard_bounds,
+        q=1,
+        num_restarts=10,
+        raw_samples=512,
+        options={"batch_limit": 5, "maxiter": 200},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+@experimental("2.4.0")
+def qehvi_candidates_func(
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+) -> "torch.Tensor":
+    """Quasi MC-based batch Expected Hypervolume Improvement (qEHVI).
+
+    The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
+    with multi-objective optimization.
+
+    .. seealso::
+        :func:`~optuna.integration.botorch.qei_candidates_func` for argument and return value
+        descriptions.
+    """
+
+    n_outcomes = train_obj.size(-1)
+
+    if train_con is not None:
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        partitioning_y = train_obj[is_feas]
+
+        constraints = []
+        n_contraints = train_con.size(1)
+
+        for i in range(n_contraints):
+            constraints.append(lambda Z, i=i: Z[..., -n_contraints + i])
+        additional_qehvi_kwargs = {
+            "objective": IdentityMCMultiOutputObjective(outcomes=list(range(n_outcomes))),
+            "constraints": constraints,
+        }
+    else:
+        train_y = train_obj
+
+        partitioning_y = train_obj
+
+        additional_qehvi_kwargs = {}
+
+    train_x = normalize(train_x, bounds=bounds)
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.shape[-1]))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_model(mll)
+
+    partitioning = NondominatedPartitioning(num_outcomes=n_outcomes, Y=partitioning_y)
```

Comment:
According to the [document](https://botorch.org/api/utils.html#botorch.utils.multi_objective.box_decomposition.NondominatedPartitioning),
> Note: it is only feasible to use this algorithm to compute an exact decomposition of the non-dominated space for m<5 objectives (alpha=0.0). 

How about adding a note: `The number of objectives should be smaller than 5 due to botorch.utils.multi_objective.box_decomposition.NondominatedPartitioning` or a TODO comment here?

Reply:
I would recommend defaulting to qParEGO instead of qEHVI for any more than 3 objectives, otherwise it can get very slow on a CPU (particularly if you aren't using approximate hypervolume).

Reply:
We have a heuristic thatfor selecting the approximation level for the partitioning in Ax that may be of interest: https://github.com/facebook/Ax/blob/master/ax/models/torch/botorch_moo_defaults.py#L303 The heuristic is based on the optimization performance and wall time results in our paper (Sec F.4): https://arxiv.org/abs/2006.05078

Reply:
The heuristic for alpha in Ax looks very practical. Thanks for the advice. As for qParEGO, I actually skipped it on purpose to keep this PR reasonably small, but given that it's pointed out by the author and @HideakiImamura, I should perhaps rethink that. Just have to run some more benchmarks to make sure it runs as expected, but let me work on those changes. Again, thanks a lot for your insightful comments.